probe-rs/src/memory/mod.rs

```diff
@@ -2,8 +2,24 @@ use crate::error::Error;
 
 use scroll::Pread;
 
+/// Call to {0} with data.len() not a multiple of {1}
+#[derive(Debug, thiserror::Error, docsplay::Display)]
+pub struct InvalidDataLengthError(&'static str, usize);
+
+/// Memory access to address {address:#X?} was not aligned to {alignment} bytes.
+#[derive(Debug, thiserror::Error, docsplay::Display)]
+pub struct MemoryNotAlignedError {
+    /// The address of the register.
+    pub address: u64,
+    /// The required alignment in bytes (address increments).
+    pub alignment: usize,
+}
+
 /// An interface to be implemented for drivers that allow target memory access.
 pub trait MemoryInterface {
+    /// The error type used accross this implementation.
+    type Error: std::error::Error + From<InvalidDataLengthError> + From<MemoryNotAlignedError>;
```

Comment:
Can we somehow get rid of this associated type? This makes everything just that much more complicated.

Reply:
Ideally what we’d want for that is a trait alias really but they are not stable yet.

I tried making a `MemoryInterface<Error = Error>` a supertrait of a new `GMemoryInterface` with a blanket impl on a `MemoryInterface<Error = Error>` bound but then it had an even bigger impact over all the dyn traits used everywhere.

Factorizing this helped fix a bug in the write function. There are other places where I think some factorization could help (eg around the Mocks) but I didn’t have time to look deeper in to this.

IMHO, although it makes it ugly (I 100% agree on that) it is still the least bad solution (vs code duplication + bugs + maintenance surface).

Reply:
I understand the reason behind this change (although what the bug was, got absolutely lost in this PR), but have you considered using a generic trait instead of an associated type?

Reply:
I didn’t try adding a generic parameter to the trait as it seemed like it would’nt make much sense to `impl MemoryInterface<ArmError> for Whatever` and `impl MemoryInterface<SomeOtherError> for Whatever`.

I tried:
```rust
trait NewMemoryInterface: … {
  type Error: …;
  …
}
trait MemoryInterface: NewMemoryInterface<Error = Error> {}
impl<T> MemoryInterface for T
where
  T: NewMemoryInterface<Error = Error> {
}
```

But this cause issues in places `dyn MemoryInterface` needs to be converted in `dyn NewMemoryInterface<Error = Error>`.

I also tried keeping `MemoryInterface` as it is, using `crate::Error`, and replace the `ArmMemoryInterface` which uses ArmError but updating this use of ArmError everywhere proved to be even more painful 😭

Reply:
The fix to the bug is visible on line 258 to 305 of `probe-rs/src/memory/mod.rs` (of the version in this PR)

Reply:
> I didn’t try adding a generic parameter to the trait as it seemed like it would’nt make much sense to impl MemoryInterface<ArmError> for Whatever and impl MemoryInterface<SomeOtherError> for Whatever.

Would you mind if I tried my idea on top of all this?

Reply:
Well I went ahead, and the result is https://github.com/probe-rs/probe-rs/commit/d0b3a7f9414661fd88cbcfaf1ea70006d58252f3 which I think is a lot nicer. On the other hand, this didn't bring the diff-reduction I was hoping for.

Reply:
I don’t mind at all, and you’ve been much faster than me 😄 
I agree it looks much nicer.
Do you mind if I update this PR to avoid introducing that noise ?
Or do you prefer me to cherry-pick your commit?

Reply:
Well, it was more tedious than I thought but this should now be addressed :)